src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1063 @@
+        ///   any IP address ("/ip4/0.0.0.0" or "/ip6/::") then all network interfaces addresses
+        ///   are added.  If the port is zero (as in "/ip6/::/tcp/0"), then the peer addresses contains the actual port number
+        ///   that was assigned.
+        ///   </para>
+        /// </remarks>
+        public Task<MultiAddress> StartListeningAsync(MultiAddress address)
+        {
+            var cancel = new CancellationTokenSource();
+
+            if (!listeners.TryAdd(address, cancel)) throw new Exception($"Already listening on '{address}'.");
+
+            // Start a listener for the transport
+            var didSomething = false;
+            foreach (var protocol in address.Protocols)
+                if (TransportRegistry.Transports.TryGetValue(protocol.Name, out var transport))
+                {
+                    address = transport().Listen(address, OnRemoteConnect, cancel.Token);
+                    listeners.TryAdd(address, cancel);
+                    didSomething = true;
+                    break;
+                }
+
+            if (!didSomething)
+                throw new ArgumentException($"Missing a transport protocol name '{address}'.", "address");
+
+            var result = new MultiAddress($"{address}/ipfs/{LocalPeer.Id}");
+
+            // Get the actual IP address(es).
+            IEnumerable<MultiAddress> addresses = new List<MultiAddress>();
+            var ips = NetworkInterface.GetAllNetworkInterfaces()
+
+                // It appears that the loopback adapter is not UP on Ubuntu 14.04.5 LTS
+               .Where(nic => nic.OperationalStatus == OperationalStatus.Up
+                 || nic.NetworkInterfaceType == NetworkInterfaceType.Loopback)
+               .SelectMany(nic => nic.GetIPProperties().UnicastAddresses);
+            if (result.ToString().StartsWith("/ip4/0.0.0.0/"))
+                addresses = ips
+                   .Where(ip => ip.Address.AddressFamily == AddressFamily.InterNetwork)
+                   .Select(ip =>
+                    {
+                        return new MultiAddress(result.ToString().Replace("0.0.0.0", ip.Address.ToString()));
+                    })
+                   .ToArray();
+            else if (result.ToString().StartsWith("/ip6/::/"))
+                addresses = ips
+                   .Where(ip => ip.Address.AddressFamily == AddressFamily.InterNetworkV6)
+                   .Select(ip => { return new MultiAddress(result.ToString().Replace("::", ip.Address.ToString())); })
+                   .ToArray();
+            else
+                addresses = new MultiAddress[] {result};
+            if (addresses.Count() == 0)
+            {
+                var msg = "Cannot determine address(es) for " + result;
+                foreach (var ip in ips) msg += " nic-ip: " + ip.Address.ToString();
+                cancel.Cancel();
+                throw new Exception(msg);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)